### ISSUE TITLE
Add e2e testing support for Windows in flutter/plugins

### ISSUE THREAD

**Issue by stuartmorgan** [flutter/flutter#56143]
In order to start landing Windows plugins in flutter/plugins, we need end-to-end test support for Windows. The `plugin_tools` support already landed, but we need CI support. That means we'll need a Docker image that can build Flutter applications for Windows—the existing images we have elsewhere in Flutter repo CIs are using 2017 so can't be reused—and the Cirrus configuration to run the tests with it.

**Comment by stuartmorgan** [comment]
Updating assignment to reflect that this is currently blocked on steps that need to be taken on the infra side.

**Comment by godofredoc** [comment]
Created the request to get 2 bots with vs2019: fxb/56788

**Comment by godofredoc** [comment]
The two new windows bots with vs219 are ready to use, next steps are to create a new recipe for plugins. Can you please share the commands we need to run after a checkout to run a test?

**Comment by godofredoc** [comment]
Please let me know if you prefer to create the recipe yourself and I'll add instructions for that in the infra playbook.

**Comment by stuartmorgan** [comment]
The steps should be:
1. Ensure that the latest version of Flutter master is checked out and in the path (hopefully we have recipe helpers for that already?)
2. `flutter config --enable-windows-desktop`
3. `pub global activate flutter_plugin_tools`
4. Check out `flutter/plugins`
5. Run the following in the root of the `flutter/plugins` checkout:
   ```
   ./script/incremental_build.sh build-examples --windows
   ./script/incremental_build.sh drive-examples --windows
   ```

It's possible that there will be a step 0 of adding more components to the VS installation, depending on what's already there, but hopefully not.

If you could do a first pass of recipe creation since my recipe experience is very limited, that would be great, and then if there are issues that need to be ironed out I can fine tune it.

**Comment by godofredoc** [comment]
Sounds good, I'll start working on this and try to get some data before the end of the week

**Comment by stuartmorgan** [comment]
Any update on this? If you aren't going to have time and can point me to instructions, I can look next week.

**Comment by godofredoc** [comment]
I advanced a little bit but got blocked:

https://chromium-swarm.appspot.com/task?id=4e2778c90603cd10

We'll need .bat or .ps shell equivalents to the bash scripts. Please reassign the bug back to me when they're ready

**Comment by stuartmorgan** [comment]
> We'll need .bat or .ps shell equivalents to the bash scripts.

Or Dart: https://github.com/flutter/flutter/issues/64025

**Comment by godofredoc** [comment]
Actually after you told me about bash being available in windows I gave it a try and it seems it is available. Taking this bug back.

https://chromium-swarm.appspot.com/task?id=4e27a8abd591be10

**Comment by godofredoc** [comment]
I made some changes to the bash script to use pub.bat instead of pub and the build completed successfully:

https://chromium-swarm.appspot.com/task?d=true&id=4e28101c250a3e10

This is the PR I'm using to modify the build script:

https://github.com/flutter/plugins/pull/2951

Even though the build was successful it didn't build any plugin, seems like none of the plugins supports windows:

```
Checking for changed packages from 0c76383c41096b1bcfd78abd9c0fdc1b6a78ac87
 No changes detected in packages.
 No changes detected in packages.
 Running for all packages
 BUILDING Windows for android_alarm_manager\example
 Windows is not supported by this plugin
 BUILDING Windows for android_intent\example
 Windows is not supported by this plugin
...
```

**Comment by stuartmorgan** [comment]
@godofredoc Now that the recipe has landed is there an easy way for me to test PRs on that bot?

**Comment by godofredoc** [comment]
You'll need to run two steps:

```
led get-builder "luci.flutter.try:Windows framework_tests" | led edit -r plugins/plugins | led edit -d os=Windows | led edit -d id=flutter-try-windows-us-central1-c-0-h9xx > properties.json
```

You'll need to open properties.json and delete these lines(this is because the new bot have different properties):
```
{
                "key": "cores",
                "value": "8",
                "expiration": "21600s"
              },
              {
                "key": "cpu",
                "value": "x64",
                "expiration": "21600s"
              },
```

Finally:

```
cat properties.json | led edit-system -p 20 | led edit -pa git_url=https://github.com/flutter/plugins | led edit -pa git_ref=refs/pull/<your_pr>/head | led launch
```

Once we have a basic plugin I can add an infra config that will simplify this process to a single command.

**Comment by stuartmorgan** [comment]
> You'll need to open properties.json and delete these lines

Are there other changes needed to that file? My [first attempt at following these instructions](https://chromium-swarm.appspot.com/task?id=4e632dfed047a610) says "0 bots could possibly run this task".

**Comment by stuartmorgan** [comment]
Poking at the bot UI linked from the "0 bots" it seems like `id:flutter-try-windows-us-central1-c-0-h9xx` is problematic. Should I be using a different ID instead?

**Comment by godofredoc** [comment]
We enabled scheduling and the bots are regenerated every day. You can use this link to identify the ones available during the day: https://chromium-swarm.appspot.com/botlist?c=id&c=task&c=os&c=status&d=asc&f=pool%3Aluci.flutter.try&f=os%3AWindows-Server&k=os&s=id

**Comment by stuartmorgan** [comment]
Thanks, testing now. I forgot to apply your incremental_build.sh changes to the first run so that failed, but `doctor` is showing that the VS checks passed, meaning that the bot should have everything we need to do builds, so one hurdle down!

**Comment by stuartmorgan** [comment]
I discovered that there was a dependency collision in the PR I was testing with, such that `pub get` on the example app (where it was hanging) would fail. I fixed that and re-ran, and got [a green run](https://luci-milo.appspot.com/raw/build/logs.chromium.org/flutter/led/stuartmorgan_google.com/8262c62e8def8247edb8e1e393a0f00077b60a9555dc8c7259f30762466019fb/+/annotations) with the test PR. I'd like to understand why it appeared to be hanging forever rather than outputting failure information, but that's no longer a blocker for progress.

What's the next step in getting these bots+recipe wired into the flutter/plugins CI?

**Comment by godofredoc** [comment]
This is great news. @keyonghan can help adding the configurations:

- [ ] PR to add to repo to cocoon
- [ ] PR to add checks validations for the repo
- [ ] Plugins repo enabled in flutter-dashboard github configs.
- [ ] flutter/infra change add a new file for plugins repo.
- [ ] add a try_configs.json file to dev/ in plugins repo.

**Comment by godofredoc** [comment]
Also @keyonghan can you please update the docs with the examples. 
https://g3doc.corp.google.com/company/teams/flutter/luci/new_repo.md?cl=head

**Comment by keyonghan** [comment]
@godofredoc Do we plan to mirror the plugin repo to GoB: https://chromium.googlesource.com/external/github.com/flutter, or we are interested in only pre builds?

**Comment by godofredoc** [comment]
I think it is presubmit only for now but let's open the bug to get the mirror just in case is needed in the future.

**Comment by keyonghan** [comment]
> I think it is presubmit only for now but let's open the bug to get the mirror just in case is needed in the future.

https://github.com/flutter/flutter/issues/65123

**Comment by keyonghan** [comment]
@godofredoc @stuartmorgan Try running the recipes, and it works by using one command `led get-builder "luci.flutter.try:Windows Plugins" | led edit-recipe-bundle | led launch`. But it uses default `os=Windows-10`. Shall we update the os to use `Windows-server`?

**Comment by stuartmorgan** [comment]
> But it uses default os=Windows-10. Shall we update the os to use Windows-server?

I don't know how that field interacts with the actual image that we are using, so I'm not the right person to answer that.

**Comment by godofredoc** [comment]
Does it work with those machines? can you please share the link to the builder run?

**Comment by keyonghan** [comment]
https://luci-milo.appspot.com/raw/build/logs.chromium.org/flutter/led/keyonghan_google.com/9015ff8121c65bd301e9b0e8cf3e20d612dbda485bb991567ec08dc27b727b10/+/annotations

**Comment by keyonghan** [comment]
For [`appInstallation`](https://github.com/flutter/cocoon/blob/master/app_dart/lib/src/datastore/cocoon_config.dart#L256), what is needed from `plugins` side? I guess we need Ray help set up? @godofredoc

**Comment by keyonghan** [comment]
Oh, never mind. I just need to add an entry to `githubapp_installations` in config.

**Comment by godofredoc** [comment]
We need help from Ray to add plugins to the Flutter-dashboard supported repos and also to add fluttergithubbot as a writer to the repo to ensure status and checks can be updated.

**Comment by keyonghan** [comment]
Have had Ray helped add `webhook` connecting to cocoon github webhook API from `plugins` repo; add `fluttergithubbot` to `plugins` repo.
And have added an entry `flutter/plugins` to `githubapp_installations` in datastore Cocoon_config table.

Is there anything else we need to do? Have we covered what you said about `help from Ray to add plugins to the Flutter-dashboard supported repos`?

**Comment by godofredoc** [comment]
Was that the flutter dashboard app? the webhook needs to be disabled.

**Comment by keyonghan** [comment]
So this step go/flutter-infra-new-repo#webhook is no longer needed?

**Comment by godofredoc** [comment]
That is correct, we need to remove it from the docs.

**Comment by keyonghan** [comment]
Thanks. Just had Ray helped remove webhook, and setup the plugin to flutter-dashboard app. 
I think we have everything ready, I will push a new deployment and give it a shot.

**Comment by keyonghan** [comment]
Builder is now successfully running in pre-submit tasks.
One test PR: https://github.com/flutter/plugins/pull/2997

**Comment by godofredoc** [comment]
This is great news, thanks Keyong!

**Comment by godofredoc** [comment]
@stuartmorgan please let us know what else will be needed.

**Comment by stuartmorgan** [comment]
Fantastic! That should be all we need at this point. I'll see if we can get a Windows plugin landed next week so we're fully exercising this new codepath regularly. Thanks for all the work getting this set up.

**Comment by stuartmorgan** [comment]
This is no longer working, actually; runs are now showing that [VS is not installed](https://logs.chromium.org/logs/flutter/buildbucket/cr-buildbucket.appspot.com/8870044690402542352/+/steps/prepare_environment/0/steps/flutter_doctor/0/stdout).

This appears to be due to the switch from Windows 10 to Windows Server, since [it was showing that doctor error in that test run](https://github.com/flutter/flutter/issues/56143#issuecomment-686721279), but previously worked. I'm guessing that that change meant we were using a completely different Windows image that doesn't include the toolchain.

**Comment by keyonghan** [comment]
Thanks @stuartmorgan for catching the issue. After we switched OS to `Windows-Server` with 32 cores, the test run passes: https://ci.chromium.org/p/flutter/builders/try/Windows%20Plugins/44

**Comment by stuartmorgan** [comment]
Thanks for the quick fix!
